xscen/diagnostics.py

```diff
@@ -2,110 +2,290 @@
+        # Create cf_time objects to compare the dates
+        end_date = xr.cftime_range(
+            start=end_date, periods=1, freq="D", calendar=ds.time.dt.calendar
+        )[0]
+
+        if not ((ds_start < end_date) and (ds_end >= end_date)):
+            err = f"The end date is not at least {end_date}. Received {ds.time.max().values.astype('datetime64[m]')}."
+            if "end_date" in raise_on:
+                raise ValueError(err)
+            else:
+                warnings.warn(err, UserWarning, stacklevel=1)
+
+    # Check variables
+    if variables_and_units is not None:
+        for v in variables_and_units:
+            if v not in ds:
+                raise ValueError(f"The variable '{v}' is missing.")
+            if ds[v].attrs.get("units", None) != variables_and_units[v]:
+                xc.core.units.check_units(
+                    ds[v], variables_and_units[v]
+                )  # Will always raise an error if the units are not compatible
+                err = f"The variable '{v}' does not have the expected units '{variables_and_units[v]}'. Received '{ds[v].attrs['units']}'."
+                if "variables_and_units" in raise_on:
+                    raise ValueError(err)
+                else:
+                    warnings.warn(err, UserWarning, stacklevel=1)
+
+    # Check CF conventions
+    if cfchecks is not None:
+        cfchecks = deepcopy(cfchecks)
+        for v in cfchecks:
+            for check in cfchecks[v]:
+                if check == "check_valid":
+                    cfchecks[v][check]["var"] = ds[v]
+                elif check == "cfcheck_from_name":
+                    cfchecks[v][check].setdefault("varname", v)
+                    cfchecks[v][check]["vardata"] = ds[v]
+                else:
+                    raise ValueError(f"Check '{check}' is not in xclim.")
+                if "cfchecks" in raise_on:
+                    warnings.simplefilter("error")
+                    try:
+                        getattr(xc.core.cfchecks, check)(**cfchecks[v][check])
+                    except UserWarning as e:
+                        raise ValueError(e)
+                    warnings.resetwarnings()
+                else:
+                    getattr(xc.core.cfchecks, check)(**cfchecks[v][check])
+
+    inferred_freq = xr.infer_freq(ds.time)
+    if freq is not None:
+        if inferred_freq is None:
+            err = "The timesteps are irregular or cannot be inferred by xarray."
+            if "freq" in raise_on:
+                raise ValueError(err)
+            else:
+                warnings.warn(err, UserWarning, stacklevel=1)
+        elif freq != inferred_freq:
+            err = f"The frequency is not '{freq}'. Received '{inferred_freq}'."
+            if "freq" in raise_on:
+                raise ValueError(err)
+            else:
+                warnings.warn(err, UserWarning, stacklevel=1)
+
+    if missing is not None:
+        if inferred_freq not in ["M", "MS", "D", "H"]:
```

Comment:
Inferring freq is not very expensive.
```suggestion
    if missing is not None:
        inferred_freq = xr.infer_freq(ds.time)
        if inferred_freq not in ["M", "MS", "D", "H"]:
```